src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,575 @@
+                num_workers=len(self.servers),
+                servers=self.servers,
+                num_requests=fastapi_app.num_current_requests,
+                processing_time=fastapi_app.last_processing_time,
+                global_request_count=fastapi_app.global_request_count,
+            )
+
+        @fastapi_app.put("/system/update-servers")
+        async def update_servers(servers: List[str], authenticated: bool = Depends(authenticate_private_endpoint)):
+            self.servers = servers
+            self._iter = cycle(self.servers)
+
+        @fastapi_app.post("/api/predict", response_model=self._output_type)
+        async def balance_api(inputs: self._input_type):
+            return await self.process_request(inputs)
+
+        uvicorn.run(
+            fastapi_app,
+            host=self.host,
+            port=self.port,
+            loop="uvloop",
+            timeout_keep_alive=self._timeout_keep_alive,
+            access_log=False,
+        )
+
+    def update_servers(self, server_works: List[LightningWork]):
+        """Updates works that load balancer distributes requests to.
+
+        AutoScaler uses this method to increase/decrease the number of works.
+        """
+        old_servers = set(self.servers)
+        server_urls: List[str] = [server.url for server in server_works if server.url]
+        new_servers = set(server_urls)
+
+        if new_servers == old_servers:
+            logger.debug("no new server added")
+            return
+
+        if new_servers - old_servers:
+            logger.info(f"servers added: {new_servers - old_servers}")
+
+        deleted_servers = old_servers - new_servers
+        if deleted_servers:
+            logger.info(f"servers deleted: {deleted_servers}")
+
+        self.send_request_to_update_servers(server_urls)
+
+    def send_request_to_update_servers(self, servers: List[str]):
+        AUTHORIZATION_TYPE = "Basic"
+        USERNAME = "lightning"
+        AUTO_SCALER_AUTH_PASSWORD = os.environ.get("AUTO_SCALER_AUTH_PASSWORD", "")
+
+        try:
+            param = f"{USERNAME}:{AUTO_SCALER_AUTH_PASSWORD}".encode()
+            data = b64encode(param).decode("utf-8")
+        except (ValueError, UnicodeDecodeError) as e:
+            raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                detail="Invalid authentication credentials",
+                headers={"WWW-Authenticate": "Basic"},
+            ) from e
+        headers = {
+            "accept": "application/json",
+            "username": USERNAME,
+            "Authorization": AUTHORIZATION_TYPE + " " + data,
+        }
+        response = requests.put(f"{self.url}/system/update-servers", json=servers, headers=headers, timeout=10)
+        response.raise_for_status()
+
+
+class AutoScaler(LightningFlow):
+    """A LightningFlow component that handles all the servers and uses load balancer to spawn up and shutdown based
+    on current requests in the queue.
```

Comment:
Maybe market this a bit more:
```suggestion
    """The ``AutoScaler`` can be used to automatically change the number of replicas of the given server in response to changes in the number of incoming requests. Incoming requests will be batched and balanced across the replicas.
```